flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java

```diff
@@ -163,8 +164,15 @@ YarnLocalResourceDescriptor registerSingleLocalResource(
         addToRemotePaths(whetherToAddToRemotePaths, resourcePath);
 
         if (Utils.isRemotePath(resourcePath.toString())) {
-            final FileStatus fileStatus = fileSystem.getFileStatus(resourcePath);
-            LOG.debug("Using remote file {} to register local resource", fileStatus.getPath());
+            final FileSystem srcFs = resourcePath.getFileSystem(fileSystem.getConf());
+            final FileStatus resourceFileStatus = srcFs.getFileStatus(resourcePath);
+            LOG.debug("Using remote file {} to register local resource", resourceFileStatus.getPath());
+            final Path applicationDir = getApplicationDirPath(homeDir, applicationId);
+            final String suffix = (relativeDstPath.isEmpty() ? "" : relativeDstPath + "/") + resourcePath.getName();
+            final Path dst = new Path(applicationDir, suffix);
+            LOG.debug("Copying from {} to {} with replication factor {}", resourcePath, dst, fileReplication);
+            FileUtil.copy(srcFs, resourcePath, fileSystem, dst, false, fileSystem.getConf());
+            final FileStatus fileStatus = fileSystem.getFileStatus(dst);
```

Comment:
Could we unify this piece of code with `uploadLocalFileToRemote`. In `uploadLocalFileToRemote`, we support to upload things to an eventual consistent FS, for example. This is missing here. Moreover,  I think we need tests for the added functionality.

Reply:
Ok, I will put this processing logic to `uploadLocalFileToRemote`. Then is it necessary to add a flag to distinguish between local and remote files in `uploadLocalFileToRemote`? because the processing of the two is bit different.

For example,`uploadLocalFileToRemote(final Path localSrcPath, final String relativeDstPath, final boolean isLocal)`.
![image](https://user-images.githubusercontent.com/13013780/103748710-7de12180-503f-11eb-8cbe-bff3eea7fcd3.png)

And copy local/remote files to Hadoop FS in `copyToRemoteApplicationDir`, as follows:
![image](https://user-images.githubusercontent.com/13013780/103749097-019b0e00-5040-11eb-8d54-202858d8cd9f.png)

Reply:
For tests, existing case `testRegisterMultipleLocalResourcesWithRemoteFiles` can cover it in class `org.apache.flink.yarn.YarnFileStageTest`, and I can add some.

Reply:
Is `isLocal` really needed? Can't any path w/o a scheme be interpreted as a local file?

Reply:
It’s possible without `isLocal `, but in this case, it will be repeated in multiple places to judge whether the file is local or remote with `Utils.isRemotePath(resourcePath.toString())`. In fact, we only need to judge once and pass this result.

Whether it is a remote file is based on method `Utils.isRemotePath(String path)`，false(w/o schema or schema is "file://") as local file.